catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/DuplicateCommands.java

```diff
@@ -140,41 +145,132 @@
+                    rejectedExecutionHandler);
+            console.printf("Running %d threads during replication.%n", multithreaded);
+
+            if (maxMetacards > 0 && maxMetacards < batchSize) {
+                batchSize = maxMetacards;
+            }
 
-        if (queryMetacards == null || queryMetacards.isEmpty()) {
-            return 0;
-        }
+            final SourceResponse originalResponse = executorService.submit(() -> query(queryFacade,
+                    filter,
+                    queryIndex.get()))
+                    .get();
+            if (originalResponse == null) {
+                return;
+            }
+
+            final long totalHits = originalResponse.getHits();
+
+            if (totalHits <= 0) {
+                return;
+            }
+
+            // If the maxMetacards is set, restrict the totalPossible to the number of maxMetacards
+            final long totalPossible;
+            if (maxMetacards > 0 && maxMetacards <= totalHits) {
+                totalPossible = maxMetacards;
+            } else {
+                totalPossible = totalHits;
+            }
 
-        List<Metacard> createdMetacards = ingestMetacards(ingestFacade, queryMetacards);
-        int failed = queryMetacards.size() - createdMetacards.size();
-        if (failed != 0) {
-            LOGGER.info("Not all records were ingested. [{}] failed", failed);
-            failedCount.addAndGet(failed);
-            failedMetacards.addAll(subtract(queryMetacards, createdMetacards));
+            printProgressAndFlush(start, totalPossible, 0);
+            executorService.submit(() -> ingestMetacards(ingestFacade,
+                    getMetacardsFromSourceResponse(originalResponse)));
+            printProgressAndFlush(start, totalPossible, ingestedCount.get());
+
+            while (queryIndex.addAndGet(batchSize) <= totalPossible) {
+                //update batch size
+                if (totalPossible - (ingestedCount.get() + failedCount.get()) < batchSize) {
+                    batchSize = (int) totalPossible - (ingestedCount.get() + failedCount.get());
+                }
+
+                final SourceResponse response = executorService.submit(() -> query(queryFacade,
+                        filter,
+                        queryIndex.get()))
+                        .get();
+                executorService.submit(() -> ingestMetacards(ingestFacade,
+                        getMetacardsFromSourceResponse(response)));
+                printProgressAndFlush(start, totalPossible, ingestedCount.get());
+            }
+
+            executorService.shutdown();
+
+            while (!executorService.isTerminated()) {
+                try {
+                    TimeUnit.SECONDS.sleep(1);
+                } catch (InterruptedException e) {
+                    // ignore
+                }
+            }
+        } else {
+            final SourceResponse originalResponse = query(queryFacade, filter, queryIndex.get());
```

Comment:
Can we put this else clause logic into a method? It looks identical to what's being executed in the `Runnable` above.